### PR TITLE
lottie: handle roundness in path

### PR DESCRIPTION
**Pull request by mgrudzinska**
Implemented rounding of corners between
bezier curves that are straight lines.

@Issue: https://github.com/thorvg/thorvg/issues/2230

before:
before_p
beforep2

after:
after_p
afterp2

### CODE REVIEW

File: src/loaders/lottie/tvgLottieProperty.h [hunk 1]
```diff
@@ -507,16 +511,88 @@ struct LottiePathSet : LottieProperty
...
-        return operator()(frameNo, cmds, pts, transform);
+        return operator()(frameNo, cmds, pts, transform, roundness);
     }
 
     void prepare() {}
+
+private:
+    void _roundCorner(Array<PathCommand>& cmds, Array<Point>& pts, const Point& prev, const Point& curr, const Point& next, float roundness)
+    {
+        auto lenPrev = mathLength(prev - curr);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:528]: it looks like we can avoid calling sqrt() in subsequent computations.

File: src/loaders/lottie/tvgLottieProperty.h [hunk 2]
```diff
@@ -455,11 +456,14 @@ struct LottiePathSet : LottieProperty
         return (*frames)[frames->count];
     }
 
-    bool operator()(float frameNo, Array<PathCommand>& cmds, Array<Point>& pts, Matrix* transform)
+    bool operator()(float frameNo, Array<PathCommand>& cmds, Array<Point>& pts, Matrix* transform, float roundness)
     {
         if (!frames) {
-            _copy(value, cmds);
-            _copy(value, pts, transform);
+            if (roundness > ROUNDNESS_EPSILON) _handleCorners(cmds, pts, transform, roundness);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:462]: roundness can also be applied to the keyframe data. Consequently, the logic should be adjusted for additional cases as well. Refer to lines 470, 476, 484, and the final one at 510 as the interpolated result.

File: src/loaders/lottie/tvgLottieProperty.h [hunk 3]
```diff
@@ -507,16 +511,88 @@ struct LottiePathSet : LottieProperty
...
+            if (exps->result<LottiePathSet>(frameNo, cmds, pts, transform, roundness, exp)) return true;
         }
-        return operator()(frameNo, cmds, pts, transform);
+        return operator()(frameNo, cmds, pts, transform, roundness);
     }
 
     void prepare() {}
+
+private:
+    void _roundCorner(Array<PathCommand>& cmds, Array<Point>& pts, const Point& prev, const Point& curr, const Point& next, float roundness)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:526]: Remove '_' for method names.

File: src/loaders/lottie/tvgLottieProperty.h [hunk 4]
```diff
@@ -507,16 +511,88 @@ struct LottiePathSet : LottieProperty
...
+
+        pts.push(curr - 2.0f * dPrev);
+        pts.push(curr - dPrev);
+        pts.push(curr - dNext);
+        pts.push(curr - 2.0f * dNext);
+        cmds.push(PathCommand::LineTo);
+        cmds.push(PathCommand::CubicTo);
+    }
+
+    void _handleCorners(Array<PathCommand>& cmds, Array<Point>& pts, Matrix* transform, float roundness)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:544]: We might need to compare the binary size where these functions(handleCorner, roundCorner) are inside the cpp file vs header file.
`size` will help you the diff.

**Review by hermet — COMMENTED** [review]
thanks please check comments.

**Comment by hermet** [comment]
could you please remove this code in this patch?
image

File: src/loaders/lottie/tvgLottieProperty.h [hunk 5]
```diff
@@ -30,6 +30,7 @@
 #include "tvgLottieInterpolator.h"
 #include "tvgLottieExpressions.h"
 
+#define ROUNDNESS_EPSILON 1.0f
```
[the remark this answers is not in the document]
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieProperty.h:33]: now this commit is in https://github.com/thorvg/thorvg/pull/2267 so ROUNDNESS_EPSILON is only in one place

File: src/loaders/lottie/tvgLottieProperty.h (hunk 2 above)
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieProperty.h:462]: ofc

File: src/loaders/lottie/tvgLottieProperty.h (hunk 3 above)
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieProperty.h:526]: I changed them to be statics in .h file, so I kept '_'. same as other internal fun

File: src/loaders/lottie/tvgLottieProperty.h (hunk 1 above)
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieProperty.h:528]: please check now

File: src/loaders/lottie/tvgLottieProperty.h (hunk 4 above)
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieProperty.h:544]: moved as statics outside the class

**Comment by mgrudzinska** [comment]
commit moved to https://github.com/thorvg/thorvg/pull/2267
all changes applied there

**Pull request closed without merging**